File: src/main/java/dev/hoot/api/packets/GroundItemPackets.java
package dev.hoot.api.packets;

import dev.hoot.api.game.Game;
import meteor.api.packets.ClientPackets;
import net.runelite.api.Client;
import net.runelite.api.TileItem;
import net.runelite.api.packets.ClientPacket;
import net.runelite.api.packets.PacketBufferNode;

import java.util.List;

public class GroundItemPackets
{
	public static void groundItemFirstOption(TileItem item, boolean ctrlDown)
	{
		queueGroundItemAction1Packet(item.getId(), item.getWorldLocation().getX(), item.getWorldLocation().getY(), ctrlDown);
	}

	public static void groundItemSecondOption(TileItem item, boolean ctrlDown)
	{
		queueGroundItemAction2Packet(item.getId(), item.getWorldLocation().getX(), item.getWorldLocation().getY(), ctrlDown);
	}

	public static void groundItemThirdOption(TileItem item, boolean ctrlDown)
	{
		queueGroundItemAction3Packet(item.getId(), item.getWorldLocation().getX(), item.getWorldLocation().getY(), ctrlDown);
	}

	public static void groundItemFourthOption(TileItem item, boolean ctrlDown)
	{
		queueGroundItemAction4Packet(item.getId(), item.getWorldLocation().getX(), item.getWorldLocation().getY(), ctrlDown);
	}

	public static void groundItemFifthOption(TileItem item, boolean ctrlDown)
	{
		queueGroundItemAction5Packet(item.getId(), item.getWorldLocation().getX(), item.getWorldLocation().getY(), ctrlDown);
	}

	public static void groundItemAction(TileItem item, String action, boolean ctrlDown)
	{
		List<String> actions = item.getActions();
		int index = actions.indexOf(action);
		switch (index)
		{
			case 0:
				groundItemFirstOption(item, ctrlDown);
				break;
			case 1:
				groundItemSecondOption(item, ctrlDown);
				break;
			case 2:
				groundItemThirdOption(item, ctrlDown);
				break;
			case 3:
				groundItemFourthOption(item, ctrlDown);
				break;
			case 4:
				groundItemFifthOption(item, ctrlDown);
				break;
		}
	}

	public static void queueItemUseOnGroundObjectPacket(int groundItemId, int worldPointX, int worldPointY, int itemSlot, int itemId, int itemWidgetId, boolean ctrlDown)
	{
		createItemOnGroundItem(groundItemId, worldPointX, worldPointY, itemSlot, itemId, itemWidgetId, ctrlDown).send();
	}

	public static void queueSpellOnGroundObjectPacket(int groundItemId, int worldPointX, int worldPointY, int spellWidgetId, boolean ctrlDown)
	{
		createSpellOnGroundItem(groundItemId, worldPointX, worldPointY, spellWidgetId, ctrlDown).send();
	}

	public static void queueGroundItemAction1Packet(int groundItemId, int worldPointX, int worldPointY, boolean ctrlDown)
	{
		createFirstAction(groundItemId, worldPointX, worldPointY, ctrlDown).send();
	}

	public static void queueGroundItemAction2Packet(int groundItemId, int worldPointX, int worldPointY, boolean ctrlDown)
	{
		createSecondAction(groundItemId, worldPointX, worldPointY, ctrlDown).send();
	}

	public static void queueGroundItemAction3Packet(int groundItemId, int worldPointX, int worldPointY, boolean ctrlDown)
	{
		createThirdAction(groundItemId, worldPointX, worldPointY, ctrlDown).send();
	}

	public static void queueGroundItemAction4Packet(int groundItemId, int worldPointX, int worldPointY, boolean ctrlDown)
	{
		createFourthAction(groundItemId, worldPointX, worldPointY, ctrlDown).send();
	}

	public static void queueGroundItemAction5Packet(int groundItemId, int worldPointX, int worldPointY, boolean ctrlDown)
	{
		createFifthAction(groundItemId, worldPointX, worldPointY, ctrlDown).send();
	}

	public static PacketBufferNode createItemOnGroundItem(int groundItemId, int worldPointX, int worldPointY,
														 int itemSlot, int itemId, int itemWidgetId, boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createItemOnGroundItemPacket("OPOBJU", groundItemId, worldPointX, worldPointY, itemSlot, itemId, itemWidgetId, ctrlDown);
	}

	public static PacketBufferNode createSpellOnGroundItem(int groundItemId, int worldPointX, int worldPointY,
												   int spellWidgetId, boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createSpellOnGroundItemPacket("OPOBJT", groundItemId, worldPointX, worldPointY, spellWidgetId, ctrlDown);
	}

	public static PacketBufferNode createFirstAction(int groundItemId, int worldPointX, int worldPointY,
													boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createGroundItemActionPacket("OPOBJ1", groundItemId, worldPointX, worldPointY, ctrlDown);
	}

	public static PacketBufferNode createSecondAction(int groundItemId, int worldPointX, int worldPointY,
													boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createGroundItemActionPacket("OPOBJ2", groundItemId, worldPointX, worldPointY, ctrlDown);
	}

	public static PacketBufferNode createThirdAction(int groundItemId, int worldPointX, int worldPointY,
													boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createGroundItemActionPacket("OPOBJ3", groundItemId, worldPointX, worldPointY, ctrlDown);
	}

	public static PacketBufferNode createFourthAction(int groundItemId, int worldPointX, int worldPointY,
													boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createGroundItemActionPacket("OPOBJ4", groundItemId, worldPointX, worldPointY, ctrlDown);
	}

	public static PacketBufferNode createFifthAction(int groundItemId, int worldPointX, int worldPointY,
													boolean ctrlDown)
	{
		return ClientPackets.INSTANCE.createGroundItemActionPacket("OPOBJ5", groundItemId, worldPointX, worldPointY, ctrlDown);
	}

	public static PacketBufferNode createWidgetOnGroundItem(int groundItemId, int worldPointX, int worldPointY, int childIndex, int itemId, int widgetId, boolean ctrlDown)
	{
		var client = Game.getClient();
		var clientPacket = Game.getClientPacket();
		PacketBufferNode packetBufferNode = Game.getClient().preparePacket(clientPacket.OPOBJT(), client.getPacketWriter().getIsaacCipher());
		packetBufferNode.getPacketBuffer().writeShort(itemId);
		packetBufferNode.getPacketBuffer().writeByte(ctrlDown ? 1 : 0);
		packetBufferNode.getPacketBuffer().writeShortLE(groundItemId);
		packetBufferNode.getPacketBuffer().writeInt(widgetId);
		packetBufferNode.getPacketBuffer().writeShort(worldPointY);
		packetBufferNode.getPacketBuffer().writeShort(childIndex);
		packetBufferNode.getPacketBuffer().writeShortAdd(worldPointX);
		return packetBufferNode;
	}
}

File: src/main/kotlin/meteor/api/packets/GroundItemPackets.kt
package meteor.api.packets

import net.runelite.api.TileItem
import net.runelite.api.packets.PacketBufferNode

object GroundItemPackets {
    fun groundItemFirstOption(item: TileItem, ctrlDown: Boolean) {
        queueGroundItemAction1Packet(
            item.id,
            item.worldLocation.x,
            item.worldLocation.y,
            ctrlDown
        )
    }

    fun groundItemSecondOption(item: TileItem, ctrlDown: Boolean) {
        queueGroundItemAction2Packet(
            item.id,
            item.worldLocation.x,
            item.worldLocation.y,
            ctrlDown
        )
    }

    fun groundItemThirdOption(item: TileItem, ctrlDown: Boolean) {
        queueGroundItemAction3Packet(
            item.id,
            item.worldLocation.x,
            item.worldLocation.y,
            ctrlDown
        )
    }

    fun groundItemFourthOption(item: TileItem, ctrlDown: Boolean) {
        queueGroundItemAction4Packet(
            item.id,
            item.worldLocation.x,
            item.worldLocation.y,
            ctrlDown
        )
    }

    fun groundItemFifthOption(item: TileItem, ctrlDown: Boolean) {
        queueGroundItemAction5Packet(
            item.id,
            item.worldLocation.x,
            item.worldLocation.y,
            ctrlDown
        )
    }

    fun groundItemAction(item: TileItem, action: String, ctrlDown: Boolean) {
        val actions: List<String> = item.actions
        val index = actions.indexOf(action)
        when (index) {
            0 -> groundItemFirstOption(item, ctrlDown)
            1 -> groundItemSecondOption(item, ctrlDown)
            2 -> groundItemThirdOption(item, ctrlDown)
            3 -> groundItemFourthOption(item, ctrlDown)
            4 -> groundItemFifthOption(item, ctrlDown)
        }
    }

    fun queueItemUseOnGroundObjectPacket(
        groundItemId: Int,
        worldPointX: Int,
        worldPointY: Int,
        itemSlot: Int,
        itemId: Int,
        itemWidgetId: Int,
        ctrlDown: Boolean
    ) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJU(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShort(worldPointY)
        packetBufferNode.packetBuffer.writeShortAddLE(itemSlot)
        packetBufferNode.packetBuffer.writeShortAddLE(groundItemId)
        packetBufferNode.packetBuffer.writeShortAdd(worldPointX)
        packetBufferNode.packetBuffer.writeIntLE(itemWidgetId)
        packetBufferNode.packetBuffer.writeByteSub(if (ctrlDown) 1 else 0)
        packetBufferNode.packetBuffer.writeShortAddLE(itemId)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueSpellOnGroundObjectPacket(
        groundItemId: Int,
        worldPointX: Int,
        worldPointY: Int,
        spellWidgetId: Int,
        ctrlDown: Boolean
    ) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJT(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeInt(spellWidgetId)
        packetBufferNode.packetBuffer.writeShortAddLE(worldPointY)
        packetBufferNode.packetBuffer.writeByteSub(if (ctrlDown) 1 else 0)
        packetBufferNode.packetBuffer.writeShortAddLE(-1)
        packetBufferNode.packetBuffer.writeShortAddLE(groundItemId)
        packetBufferNode.packetBuffer.writeShortAdd(worldPointX)
        packetBufferNode.packetBuffer.writeShortAdd(-1)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueGroundItemAction1Packet(groundItemId: Int, worldPointX: Int, worldPointY: Int, ctrlDown: Boolean) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJ1(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeByte(if (ctrlDown) 1 else 0)
        packetBufferNode.packetBuffer.writeShort(worldPointY)
        packetBufferNode.packetBuffer.writeShortLE(groundItemId)
        packetBufferNode.packetBuffer.writeShort(worldPointX)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueGroundItemAction2Packet(groundItemId: Int, worldPointX: Int, worldPointY: Int, ctrlDown: Boolean) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJ2(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShortLE(worldPointY)
        packetBufferNode.packetBuffer.writeByteAdd(if (ctrlDown) 1 else 0)
        packetBufferNode.packetBuffer.writeShortAdd(worldPointX)
        packetBufferNode.packetBuffer.writeShortAdd(groundItemId)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueGroundItemAction3Packet(groundItemId: Int, worldPointX: Int, worldPointY: Int, ctrlDown: Boolean) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJ3(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShort(worldPointY)
        packetBufferNode.packetBuffer.writeShort(groundItemId)
        packetBufferNode.packetBuffer.writeShortLE(worldPointX)
        packetBufferNode.packetBuffer.writeByte(if (ctrlDown) 1 else 0)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueGroundItemAction4Packet(groundItemId: Int, worldPointX: Int, worldPointY: Int, ctrlDown: Boolean) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJ4(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShortAddLE(groundItemId)
        packetBufferNode.packetBuffer.writeShortAdd(worldPointX)
        packetBufferNode.packetBuffer.writeShortAddLE(worldPointY)
        packetBufferNode.packetBuffer.writeByteSub(if (ctrlDown) 1 else 0)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueGroundItemAction5Packet(groundItemId: Int, worldPointX: Int, worldPointY: Int, ctrlDown: Boolean) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode: PacketBufferNode =
            client.preparePacket(clientPacket.OPOBJ5(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShortAdd(worldPointX)
        packetBufferNode.packetBuffer.writeShort(worldPointY)
        packetBufferNode.packetBuffer.writeShortLE(groundItemId)
        packetBufferNode.packetBuffer.writeByteNeg(if (ctrlDown) 1 else 0)
        client.packetWriter.queuePacket(packetBufferNode)
    }
}